### PR TITLE
[BEAM-12164] Feat: Add new restriction tracker to be able to track partition state along with timestamp for change streams connector.

### PR DESCRIPTION
**Pull request by nancyxu123**
The overall project is to remove watermark updates to the metadata table from the Connector. To do so, we need to create a new restriction tracker that keeps track of partition token state, as well as timestmap.

The purpose of this PR is to create the new restriction tracker that keeps track of both state and timestamp information.

The overall algorithm to remove watermark updates to the connector metadata table is here: https://docs.google.com/document/d/1YqpR2iiHV5ESPBg24y_BUuSKYcsgYKDhpxyNoAUzP68/edit?resourcekey=0-m4Rif7BtMjN4OxTzXbMC3g&disco=AAAAWDcj4KU

The design for the new restriction tracker is here: https://docs.google.com/document/d/1lSLOPSiJFUV2Tr7plSQXn-DcyBNgV5ELJ8p9PRpcO2c/edit?resourcekey=0-tXh0F0I9HK30XV9a7hnfiw

A previous implementation for this restriction tracker is here: https://github.com/thiagotnunes/beam/tree/spanner-cdc/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/restriction

### CODE REVIEW

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionProgressChecker.java
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionProgressChecker.java:41]: This comment is outdated

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionProgressChecker.java
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionProgressChecker.java:57]: This is not up to date? Take a look at how we are calculation progress currently in https://github.com/apache/beam/blob/df67c817184acafc4adbc56a725eed4985053171/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/TimestampRangeTracker.java#L196-L235

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionTracker.java
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionTracker.java:82]: This is already printed in the partition splitter.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionTracker.java
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionTracker.java:75]: We should also exempt the initial partition from splitting in the query change stream mode: https://github.com/apache/beam/blob/df67c817184acafc4adbc56a725eed4985053171/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/ReadChangeStreamPartitionRangeTracker.java#L77-L84

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionTracker.java
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionTracker.java:136]: I think this is a bit more nuanced, especially for the QUERY_CHANGE_STREAM mode: https://github.com/apache/beam/blob/df67c817184acafc4adbc56a725eed4985053171/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/TimestampRangeTracker.java#L173-L194

For the other modes, I am not sure if this logic is valid. We will have to test it (my fear is that a primary split is trying to finish, but it can't because this throws an exception).

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/util/TimestampGenerator.java
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/util/TimestampGenerator.java:27]: Why do we need to do this?

I think that the generator itself should produce values that are valid within its domain (MIN, MAX). If we need to restrict it, we can configure our test to only allow for values where seconds >= 0.

This way the generator is general and can be used for other use cases that need to test this.

**Review by thiagotnunes — COMMENTED** [review]
Thanks for taking care of this. Just a few thoughts:

I think that the usage of these new trackers is very risky. In the past we ran into several issues and also when using the TimestampRangeTracker approach. When testing this, we would benefit from running from a long time, with and without autoscaling and updating the pipeline midway. I know this won't be tackled in this PR per say, but just something to keep in mind.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionTracker.java
Comment (nancyxu123) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionTracker.java:40]: Hmm I think that would require significant refactoring of this PR, which might delay it even further. I think I already tried to make the QUERY_CHANGE_STREAM behavior of this new restriction tracker very similar to ReadChangeStreamPartitionRangeTracker?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionClaimer.java
Comment (nancyxu123) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionClaimer.java]: Done

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionProgressChecker.java
Comment (nancyxu123) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionProgressChecker.java]: DOne

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionProgressChecker.java
Comment (nancyxu123) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionProgressChecker.java:57]: Hmmm, so TimestampRangeTracker sets the total work as either the number of seconds till now, or the number of seconds in the total restriction right? And the current work is the number of seconds in the most recently claimed position. And the remaining work is total work - current work.

I think my restriction tracker does the same, except it includes transition modes as extra units of work?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionTracker.java
Comment (nancyxu123) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionTracker.java]: Done

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionTracker.java
Comment (nancyxu123) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionTracker.java:77]: I did do this in the PartitionRestrictionSplitter class.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionTracker.java
Comment (nancyxu123) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionTracker.java:125]: You're right. So basically, we call checkDone on the primary restriction right after the split happens right? I added the check that if the mode is QUERY_CHANGE_STREAM, we have to make sure the last claimed timestamp is the very last one in the range.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/util/TimestampGenerator.java
Comment (nancyxu123) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/util/TimestampGenerator.java]: Done

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionProgressChecker.java
Comment (thiagotnunes) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionProgressChecker.java:57]: Ah, you are doing this in the restriction tracker instead of in this class right? I think I missed that. Is there anyway we could have the full logic encapsulated in this class?

We should also include @hengfengli into the discussion, since he is the expert in this area.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionTracker.java
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionTracker.java:144]: Do I understand it correctly that for the modes != {STOP, QUERY_CHANGE_STREAM, DONE} we will always succeed if we claimed a position?

Comment (nancyxu123) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionTracker.java:144]: No....


For modes STOP/DONE, they should always succeed if we claimed a position

For mode QUERY_CHANGE_STREAM, it should only succeed if we claimed endTimestamp - 1, for restriction [startTimestamp, endTimestamp)

For all other modes, an error should be thrown, since in the splitter, I don't believe the primary restriction will ever be one of the other modes after a split.

Comment (thiagotnunes) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionTracker.java:134]: Ah ok, so when we split and we are currently in the WAIT_FOR_CHILD_PARTITIONS or in the UPDATE_STATE modes, we will update the primary restriction to be STOP, which will make the checkDone call succeed, right?

**Review by thiagotnunes — APPROVED** [review]
LGTM, I would wait on a review from @hengfengli as well before merging this to confirm the autoscaling behavior.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionTracker.java
Comment (nancyxu123) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionTracker.java:134]: Yes, exactly!

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionProgressChecker.java
Comment (hengfengli) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionProgressChecker.java:85]: Why do we need to add the total number of mode transitions here?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionProgressChecker.java
Comment (hengfengli) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionProgressChecker.java:66]: LGTM. The logic looks same to the current implementation. The only question is about why we need to add the number of transitions to the total work. It's a small issue as we are not required to have an absolute accurate progress here.

**Review by hengfengli — APPROVED** [review]
LGTM.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionProgressChecker.java
Comment (nancyxu123) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionProgressChecker.java:53]: Ah makes sense. Removed transitions from total work.

**Comment by pabloem** [comment]
LGTM thanks!

**Pull request merged**
